File: ClearScript/V8/ClearScriptV8/V8ScriptImpl.cpp
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT license.

#include "ClearScriptV8Managed.h"

namespace Microsoft {
namespace ClearScript {
namespace V8 {

    //-------------------------------------------------------------------------
    // V8ScriptImpl implementation
    //-------------------------------------------------------------------------

    V8ScriptImpl::V8ScriptImpl(ClearScript::UniqueDocumentInfo^ documentInfo, V8ScriptHolder* pHolder):
        V8Script(documentInfo, (UIntPtr)pHolder->GetCodeDigest()),
        m_gcLock(gcnew Object),
        m_pspHolder(new SharedPtr<V8ScriptHolder>(pHolder))
    {
    }

    //-------------------------------------------------------------------------

    SharedPtr<V8ScriptHolder> V8ScriptImpl::GetHolder()
    {
        BEGIN_LOCK_SCOPE(m_gcLock)

            if (m_pspHolder == nullptr)
            {
                throw gcnew ObjectDisposedException(ToString());
            }

            return *m_pspHolder;

        END_LOCK_SCOPE
    }

    //-------------------------------------------------------------------------

    V8ScriptImpl::~V8ScriptImpl()
    {
        SharedPtr<V8ScriptHolder> spHolder;

        BEGIN_LOCK_SCOPE(m_gcLock)

            if (m_pspHolder != nullptr)
            {
                // hold V8 script holder for destruction outside lock scope
                spHolder = *m_pspHolder;
                delete m_pspHolder;
                m_pspHolder = nullptr;
            }

        END_LOCK_SCOPE

        if (!spHolder.IsEmpty())
        {
            GC::SuppressFinalize(this);
        }
    }

    //-------------------------------------------------------------------------

    V8ScriptImpl::!V8ScriptImpl()
    {
        if (m_pspHolder != nullptr)
        {
            delete m_pspHolder;
            m_pspHolder = nullptr;
        }
    }

}}}

File: ClearScript/V8/SplitProxy/V8ScriptImpl.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT license.

using System;

namespace Microsoft.ClearScript.V8.SplitProxy
{
    internal sealed class V8ScriptImpl : V8.V8Script
    {
        private V8EntityHolder holder;

        public V8Script.Handle Handle => (V8Script.Handle)holder.Handle;

        public V8ScriptImpl(UniqueDocumentInfo documentInfo, UIntPtr codeDigest, V8Script.Handle hScript)
            : base(documentInfo, codeDigest)
        {
            holder = new V8EntityHolder("V8 compiled script", () => hScript);
        }

        #region disposal / finalization

        public override void Dispose()
        {
            holder.ReleaseEntity();
            GC.KeepAlive(this);
        }

        ~V8ScriptImpl()
        {
            V8EntityHolder.Destroy(ref holder);
        }

        #endregion
    }
}